File: 450-delete-node-in-a-bst/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function deleteNode(root: TreeNode | null, key: number): TreeNode | null {
    let tempRoot = new TreeNode(0, root, null);
    
    let parent = tempRoot;
    let direction = 'left';
    let deleted = root;
    while(!!deleted && deleted.val !== key) {
        parent = deleted;
        if(key < deleted.val) { deleted = deleted.left; direction = 'left'; }
        else { deleted = deleted.right; direction = 'right'; }
    }
    if(!deleted) return root;
    
    if(!deleted.left && !deleted.right) parent[direction] = null;
    else if(!deleted.left || !deleted.right) {
        parent[direction] = deleted.left || deleted.right;
    }else {
        let minimum = deleted.right;
        while(minimum.left) minimum = minimum.left;
        const val = minimum.val
        deleteNode(root, val);
        deleted.val = val;
    }
       
    return tempRoot.left;
};

File: 450-delete-node-in-a-bst/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void deleteByCases(TreeNode* node, TreeNode* parent, bool isLeft, int key) {
        if(node == NULL) return;
        if(node->val == key) {
            if(node->left == NULL && node->right == NULL) {
                if(isLeft) parent->left = NULL;
                else parent->right = NULL;
            } else if(node->left == NULL) {
                if(isLeft) parent->left = node->right;
                else parent->right = node->right;
            } else if(node->right == NULL) {
                if(isLeft) parent->left = node->left;
                else parent->right = node->left;
            } else {
                TreeNode* now = node->right;
                while(now->left != NULL) now = now->left;
                node->val = now->val;
                deleteByCases(node->right, node, false, node->val);
            }
        } else if(node->val > key) deleteByCases(node->left, node, true, key);
        else deleteByCases(node->right, node, false, key);
    }

    TreeNode* deleteNode(TreeNode* root, int key) {
        TreeNode* tmpRoot = new TreeNode(0, root, NULL);
        deleteByCases(root, tmpRoot, true, key);
        return tmpRoot->left;
    }
};
